File: src/ws-ckpt/src/plugins/hermes/cron.py
"""Crontab management for scheduled ws-ckpt snapshots."""

from __future__ import annotations

import fcntl
import os
import re
import subprocess
import tempfile
from typing import List, Optional

_LOCK_PATH = os.path.join(tempfile.gettempdir(), "ws-ckpt-cron.lock")

# Match: ws-ckpt checkpoint ... -w '<path>' or -w <path>
_CRON_RE = re.compile(r"^\S+\s+\S+\s+\S+\s+\S+\s+\S+$")
_MARKER_RE = re.compile(r"ws-ckpt\s+checkpoint\s+.*-w\s+'([^']+)'")
_MARKER_RE_UNQUOTED = re.compile(r"ws-ckpt\s+checkpoint\s+.*-w\s+(\S+)")


def _build_cron_line(workspace: str, schedule: str) -> str:
    quoted_ws = "'" + workspace.replace("'", "'\\''") + "'"
    return (
        f"{schedule} ws-ckpt checkpoint -w {quoted_ws}"
        f' -i "cron-$(date +\\%s)"'
        f' -m "scheduled snapshot"'
        f" --metadata '{{\"auto\":true,\"type\":\"cron\"}}'"
        f" >/dev/null 2>&1"
    )


def _read_crontab() -> Optional[List[str]]:
    """Return current crontab lines, or None on failure."""
    try:
        result = subprocess.run(
            ["crontab", "-l"],
            capture_output=True, text=True, timeout=10,
        )
        if result.returncode != 0:
            # "no crontab for <user>" is normal (exit 1 on first use)
            if result.returncode == 1 and "no crontab for" in (result.stderr or ""):
                return []
            return None
        return [l for l in result.stdout.splitlines() if l]
    except Exception:
        return None


def _write_crontab(lines: List[str]) -> bool:
    content = "\n".join(lines)
    if content and not content.endswith("\n"):
        content += "\n"
    try:
        result = subprocess.run(
            ["crontab", "-"],
            input=content, capture_output=True, text=True, timeout=10,
        )
        return result.returncode == 0
    except Exception:
        return False


def _extract_workspace(line: str) -> Optional[str]:
    m = _MARKER_RE.search(line)
    if m:
        return m.group(1)
    m = _MARKER_RE_UNQUOTED.search(line)
    if m:
        return m.group(1)
    return None


def _match_workspace(line: str, workspace: str) -> bool:
    return _extract_workspace(line) == workspace


def validate_cron_expr(expr: str) -> bool:
    """Return True if expr looks like a valid 5-field cron expression."""
    return bool(_CRON_RE.match(expr.strip()))


def parse_schedules_update(value: str, current: List[str]) -> tuple[Optional[List[str]], Optional[str]]:
    """Parse a cronSchedules sub-command and apply it to current list.

    Returns (new_list, None) on success or (None, error_message) on failure.
    """
    import json as _json

    sub_action, _, sub_val = value.strip().partition(" ")
    sub_action = sub_action.lower()
    sub_val = sub_val.strip()
    if len(sub_val) >= 2 and sub_val[0] == sub_val[-1] and sub_val[0] in ('"', "'"):
        sub_val = sub_val[1:-1]

    result = list(current)
    if sub_action == "add":
        if not sub_val:
            return None, "add requires a cron expression"
        if not validate_cron_expr(sub_val):
            return None, f'Invalid cron expression: "{sub_val}". Expected 5 fields, e.g. "0 * * * *"'
        if sub_val not in result:
            result.append(sub_val)
    elif sub_action == "remove":
        if not sub_val:
            return None, "remove requires a cron expression"
        if sub_val in result:
            result.remove(sub_val)
        else:
            return None, f'"{sub_val}" not found in current schedules'
    elif sub_action == "set":
        try:
            parsed = _json.loads(sub_val)
            if not isinstance(parsed, list):
                raise ValueError
            for e in parsed:
                if not validate_cron_expr(str(e)):
                    return None, f'Invalid cron expression in array: "{e}"'
            result = [str(e) for e in parsed]
        except (ValueError, _json.JSONDecodeError):
            return None, "set requires a JSON array, e.g. '[\"0 * * * *\"]'"
    else:
        return None, (
            f'Unknown cronSchedules sub-action: {sub_action}. '
            'Use: add "EXPR", remove "EXPR", or set \'["EXPR"]\''
        )
    return result, None


class CrontabManager:
    """Manage crontab entries for ws-ckpt scheduled snapshots."""

    @staticmethod
    def _with_lock(fn):
        """Execute fn while holding an exclusive flock to prevent TOCTOU races."""
        fd = os.open(_LOCK_PATH, os.O_CREAT | os.O_RDWR, 0o600)
        try:
            fcntl.flock(fd, fcntl.LOCK_EX)
            return fn()
        finally:
            fcntl.flock(fd, fcntl.LOCK_UN)
            os.close(fd)

    @staticmethod
    def sync(workspace: str, schedules: List[str]) -> bool:
        def _do():
            lines = _read_crontab()
            if lines is None:
                return False
            kept = [l for l in lines if not _match_workspace(l, workspace)]
            for s in schedules:
                kept.append(_build_cron_line(workspace, s))
            return _write_crontab(kept)
        return CrontabManager._with_lock(_do)

    @staticmethod
    def remove(workspace: str) -> bool:
        return CrontabManager.sync(workspace, [])

    @staticmethod
    def sync_with_retry(workspace: str, schedules: List[str], retries: int = 3) -> bool:
        for _ in range(retries):
            if CrontabManager.sync(workspace, schedules):
                return True
        return False

    @staticmethod
    def remove_with_retry(workspace: str, retries: int = 3) -> bool:
        return CrontabManager.sync_with_retry(workspace, [], retries)

    @staticmethod
    def migrate(old_workspace: str, new_workspace: str, schedules: List[str]) -> List[str]:
        """Remove old crontab entries and install under new workspace. Returns warnings."""
        warnings: List[str] = []
        if old_workspace and old_workspace != new_workspace:
            if not CrontabManager.remove_with_retry(old_workspace):
                warnings.append(
                    f"WARNING: Failed to remove cron entries for old workspace {old_workspace}. "
                    f"Run `crontab -e` to manually remove lines containing -w '{old_workspace}'."
                )
        if schedules:
            if not CrontabManager.sync_with_retry(new_workspace, schedules):
                warnings.append(
                    f"WARNING: Failed to install cron entries for {new_workspace}. "
                    f"Cron snapshots will not run until next session start or manual retry."
                )
        return warnings

    @staticmethod
    def list_installed(workspace: str) -> List[str]:
        lines = _read_crontab()
        if lines is None:
            return []
        result: List[str] = []
        for line in lines:
            if not _match_workspace(line, workspace):
                continue
            parts = line.split()
            if len(parts) >= 5:
                result.append(" ".join(parts[:5]))
        return result

File: src/ws-ckpt/src/plugins/openclaw/src/cron.ts
import { mkdirSync, rmdirSync } from "fs";
import { tmpdir } from "os";
import { join } from "path";
import { runCrontab } from "./commands.js";

const CRON_RE = /^\S+\s+\S+\s+\S+\s+\S+\s+\S+$/;
const LOCK_DIR = join(tmpdir(), "ws-ckpt-cron.lock");

export function validateCronExpr(expr: string): boolean {
  return CRON_RE.test(expr.trim());
}

// Match: ws-ckpt checkpoint ... -w '<path>' or -w <path>
const MARKER_RE_QUOTED = /ws-ckpt\s+checkpoint\s+.*-w\s+'([^']+)'/;
const MARKER_RE_UNQUOTED = /ws-ckpt\s+checkpoint\s+.*-w\s+(\S+)/;

function shellQuote(s: string): string {
  return "'" + s.replace(/'/g, "'\\''") + "'";
}

function buildCronLine(workspace: string, schedule: string): string {
  return (
    `${schedule} ws-ckpt checkpoint -w ${shellQuote(workspace)}` +
    ` -i "cron-$(date +\\%s)"` +
    ` -m "scheduled snapshot"` +
    ` --metadata '{"auto":true,"type":"cron"}'` +
    ` >/dev/null 2>&1`
  );
}

async function withLock<T>(fn: () => Promise<T>): Promise<T> {
  const maxWait = 5000;
  const interval = 50;
  let waited = 0;
  while (true) {
    try {
      mkdirSync(LOCK_DIR);
      break;
    } catch {
      if (waited >= maxWait) break; // proceed unlocked rather than fail
      await new Promise((r) => setTimeout(r, interval));
      waited += interval;
    }
  }
  try {
    return await fn();
  } finally {
    try { rmdirSync(LOCK_DIR); } catch { /* already removed */ }
  }
}

async function readCrontab(): Promise<string[] | null> {
  const result = await runCrontab(["-l"], { timeout: 10_000 });
  if (result.exitCode !== 0) {
    if (result.exitCode === 1 && result.stderr.includes("no crontab for")) {
      return [];
    }
    return null;
  }
  return result.stdout.split("\n").filter((l: string) => l !== "");
}

async function writeCrontab(lines: string[]): Promise<boolean> {
  let content = lines.join("\n");
  if (content && !content.endsWith("\n")) content += "\n";
  const result = await runCrontab(["-"], { input: content, timeout: 10_000 });
  return result.exitCode === 0;
}

function extractWorkspace(line: string): string | null {
  let m = MARKER_RE_QUOTED.exec(line);
  if (m) return m[1];
  m = MARKER_RE_UNQUOTED.exec(line);
  if (m) return m[1];
  return null;
}

function matchesWorkspace(line: string, workspace: string): boolean {
  return extractWorkspace(line) === workspace;
}

export type ParseResult = { schedules: string[] } | { error: string };

export function parseSchedulesUpdate(value: string, current: string[]): ParseResult {
  const spaceIdx = value.indexOf(" ");
  const subAction = (spaceIdx >= 0 ? value.slice(0, spaceIdx) : value).toLowerCase();
  let subVal = spaceIdx >= 0 ? value.slice(spaceIdx + 1).trim() : "";
  if (subVal.length >= 2 && subVal[0] === subVal[subVal.length - 1] && (subVal[0] === '"' || subVal[0] === "'")) {
    subVal = subVal.slice(1, -1);
  }

  const result = [...current];
  if (subAction === "add") {
    if (!subVal) return { error: "add requires a cron expression" };
    if (!validateCronExpr(subVal)) {
      return { error: `Invalid cron expression: "${subVal}". Expected 5 fields, e.g. "0 * * * *"` };
    }
    if (!result.includes(subVal)) result.push(subVal);
  } else if (subAction === "remove") {
    if (!subVal) return { error: "remove requires a cron expression" };
    const idx = result.indexOf(subVal);
    if (idx < 0) return { error: `"${subVal}" not found in current schedules` };
    result.splice(idx, 1);
  } else if (subAction === "set") {
    try {
      const parsed = JSON.parse(subVal);
      if (!Array.isArray(parsed)) throw new Error("not an array");
      for (const e of parsed) {
        if (!validateCronExpr(String(e))) {
          return { error: `Invalid cron expression in array: "${e}"` };
        }
      }
      return { schedules: parsed.map(String) };
    } catch {
      return { error: "set requires a JSON array, e.g. '[\"0 * * * *\"]'" };
    }
  } else {
    return { error: `Unknown cronSchedules sub-action: ${subAction}. Use: add "EXPR", remove "EXPR", or set '["EXPR"]'` };
  }
  return { schedules: result };
}

export class CrontabManager {
  static async sync(workspace: string, schedules: string[]): Promise<boolean> {
    return withLock(async () => {
      const lines = await readCrontab();
      if (lines === null) return false;
      const kept = lines.filter((l) => !matchesWorkspace(l, workspace));
      for (const s of schedules) {
        kept.push(buildCronLine(workspace, s));
      }
      return writeCrontab(kept);
    });
  }

  static async remove(workspace: string): Promise<boolean> {
    return CrontabManager.sync(workspace, []);
  }

  static async syncWithRetry(workspace: string, schedules: string[], retries = 3): Promise<boolean> {
    for (let i = 0; i < retries; i++) {
      if (await CrontabManager.sync(workspace, schedules)) return true;
    }
    return false;
  }

  static async removeWithRetry(workspace: string, retries = 3): Promise<boolean> {
    return CrontabManager.syncWithRetry(workspace, [], retries);
  }

  static async migrate(
    oldWorkspace: string,
    newWorkspace: string,
    schedules: string[],
  ): Promise<string[]> {
    const warnings: string[] = [];
    if (oldWorkspace && oldWorkspace !== newWorkspace) {
      if (!(await CrontabManager.removeWithRetry(oldWorkspace))) {
        warnings.push(
          `WARNING: Failed to remove cron entries for old workspace ${oldWorkspace}. ` +
          `Run \`crontab -e\` to manually remove lines containing -w '${oldWorkspace}'.`
        );
      }
    }
    if (schedules.length > 0) {
      if (!(await CrontabManager.syncWithRetry(newWorkspace, schedules))) {
        warnings.push(
          `WARNING: Failed to install cron entries for ${newWorkspace}. ` +
          `Cron snapshots will not run until next session start or manual retry.`
        );
      }
    }
    return warnings;
  }

  static async listInstalled(workspace: string): Promise<string[]> {
    const lines = await readCrontab();
    if (lines === null) return [];
    const result: string[] = [];
    for (const line of lines) {
      if (!matchesWorkspace(line, workspace)) continue;
      const parts = line.split(/\s+/);
      if (parts.length >= 5) result.push(parts.slice(0, 5).join(" "));
    }
    return result;
  }
}
